add mv3 info in kb (#533)
AdguardTeam/KnowledgeBase: docs/general/ad-filtering/create-own-filters.md

Apps in the modifier value cannot have a wildcard, e.g. `$app=com.*.music`.
Rules with such modifier are considered invalid.

> In [AdGuard for Chrome MV3][ext-mv3] `regexp` and `any_tld domains` are not supported.

Reviewer: it should be related to `$domain`, not `$app`

Apps in the modifier value cannot have a wildcard, e.g. `$app=com.*.music`.
Rules with such modifier are considered invalid.

> In AdGuard Browser Extension MV3, regexps and any tld domains are not supported.

Reviewer 2: "AdGuard Browser Extension MV3" or "AdGuard for Chrome MV3"?

What will we do when it's released (in a couple of weeks) and the main extension will become MV3?

Reviewer 3: @vozersky @AlinaNaletova any thoughts on the naming?

Reviewer 4: I'd say "AdGuard AdBlocker (MV3)" or even "AdGuard AdBlocker" will be fine for Chrome store

> In AdGuard Browser Extension for Chrome MV3, `regexp` and `any_tld domains` are not supported.

Reviewer 2: Products not linked

:::

:::caution Restrictions

> In AdGuard Browser Extension MV3 regexs and any tld domains are not supported.

Reviewer 2: This is documentation, it must clearly indicate when we're using technical terms.

Something like this:
```suggestion
In AdGuard Browser Extension MV3 `regexp` and `any_tld_domain` are not supported.
```

> In AdGuard Browser Extension for Chrome MV3, `regexp` and `any_tld domains` are not supported.

Reviewer 2: Products not linked

  Otherwise, page should be loaded.
  It can be combined with other request type modifiers, such as `$third-party` and `$important`.
- Rules with `$popup` modifier are not supported by AdGuard Content Blocker.

> - In AdGuard Browser Extension for Chrome MV3, rules with the [`$popup`](https://github.com/AdguardTeam/tsurlfilter/tree/epic/tswebextension/packages/tsurlfilter/src/rules/declarative-converter#popup) modifier would not work, so we disable converting them to declarative rules. We will try to use them only in our TSurlFilter engine and close new tabs programmatically.

Reviewer 2: Products not linked

:::

:::caution Restrictions

> In AdGuard Browser Extension MV3, it cannot be converted to an MV3 declarative rule, but can be implemented on the content-script side.

Reviewer 2: What? What kind of restriction is this?

@105th what is the restriction? Is it supported or not?

Reviewer 5: @SeverAnna looks like i placed old info to the table about $popup. currently we convert it - https://github.com/AdguardTeam/tsurlfilter/tree/epic/tswebextension/packages/tsurlfilter/src/rules/declarative-converter#popup. But they would not work in MV3, so we will disable converting of them.
We will try to use it only in our tsurlfilter engine and close new tabs programmatically.

<!-- Please keep them sorted -->

> | Modifier \ Products | [CoreLibs apps][cl-apps] | [AdGuard for Chromium][ext-chr] | [AdGuard for Chrome MV3][ext-mv3] | [AdGuard for Firefox][ext-ff] | [AdGuard for iOS][ios-app] | [AdGuard for Safari][ext-saf] | [AdGuard Content Blocker][and-cb] |

Reviewer 5: @SeverAnna can you please attach screenshot here (in answer to my comment)? for help checking correctness. i do not want to run locally PR.

- `@@||example.com^$jsinject` disables javascript on pages at `example.com` and all subdomains.

:::caution Restrictions

> The `$jsinject` modifier is supported, but not implemented in AdGuard Browser Extension MV3.

Reviewer 2: @105th what is this supposed to mean? It cannot be supported and not implemented, it makes literally zero sense.

Reviewer 5: these are not convertible to DNR, we use it only in our tsurlfilter engine to disable some cosmetic rules.

:::info Compatibility

> - Stealth Mode (Tracking protection) is available in AdGuard for Windows, Mac, and Android, and AdGuard Browser Extension. All other products will ignore the rules with `$stealth` modifier.

Reviewer 2: It's the other way around: "Tracking protection (prev. Stealth mode)"

:::info Compatibility

> - Stealth Mode is available in AdGuard for Windows, Mac, and Android, and AdGuard Browser Extension (but not in the MV3 version). All other products will ignore the rules with `$stealth` modifier.

Reviewer 2: Please figure out the FINAL wording that we'll be using in the future.

Will we continue to call the Chrome extension "MV3 version"?
Or will we call it Chrome extension instead?

Reviewer 2: Whenever you mention a product, it would be nice to have an anchor link to the section that explains what product it is.

Rules with `$urlblock` modifier are not supported by AdGuard Content Blocker.

> The `$urlblock` modifier is supported but not implemented in AdGuard Browser Extension MV3.

Reviewer 2: This again, "supported but not implemented" makes no sense

Reviewer 5: @SeverAnna 
so here we should delete restriction because we will implement it (someday)

| --- | :---: | :---: | :---: | :---: | :---: | :---: | :---: |
| [$all](#all-modifier) | ✅ | ✅ | ✅ [*[1]](#all-modifier-limitations) | ✅ | ✅ | ✅ | ❌ |
| [$badfilter](#badfilter-modifier) | ✅ | ✅ | ✅ [*[2]](#badfilter-modifier-limitations)| ✅ | ✅ | ✅ | ❌ |
| [$cookie](#cookie-modifier) | ✅ | ✅ | ✅ [*[3]](#cookie-modifier-limitations) | ✅ | ❌ | ❌ | ❌ |
| [$csp](#csp-modifier) | ✅ | ✅ | ✅ | ✅ | ❌ | ❌ | ❌ |
| [$hls](#hls-modifier) | ✅ | ❌ | ❌ | ❌ | ❌ | ❌ | ❌ |
| [$inline-font](#inline-font-modifier) | ✅ | ✅ | ✅ | ✅ | ❌ | ❌ | ❌ |
| [$inline-script](#inline-script-modifier) | ✅ | ✅ | ✅ | ✅ | ❌ | ❌ | ❌ |
| [$jsonprune](#jsonprune-modifier) | ✅ | ❌ | ❌ | ❌ | ❌ | ❌ | ❌ |
| [$xmlprune](#xmlprune-modifier) | ✅ | ❌ | ❌ | ❌ | ❌ | ❌ | ❌ |
| [$network](#network-modifier) | ✅ | ❌ | ❌ | ❌ | ❌ | ❌ | ❌ |
| [$permissions](#permissions-modifier) | ✅ [*[4]](#permissions-modifier-limitations) | 🧩 | 🧩 | 🧩 [*[4]](#permissions-modifier-limitations) | ❌ | ❌ | ❌ |

> | [$redirect](#redirect-modifier) | ✅ | ✅ | ✅ [*[5]](#redirect-modifier-limitations) | ✅ | ❌ | ❌ | ❌ |

Reviewer: I had to make it `*[<numver>]` because maximum number of wildcard to display correctly is 4, and for 5 of them there is only one rendered.

e.g.
```
     * -> *
    ** -> **
   *** -> ***
  **** -> ****   // ok yet
 ***** -> *      // not ok
****** -> **
```

:::info Compatibility

Rules with `$badfilter` modifier are not supported by AdGuard Content Blocker.

> In the current implementation of AdGuard Browser Extension MV3, it works across all filters, but it does not cover rules with `$domain` instersections.

Reviewer 2: @105th what is `$domain` intersections

Reviewer 5: @SeverAnna we should place here some examples to illustrate that we applied $badfilter in DNR way only if it "full cancel" source rule. if it just a partial cancelling, we cannot calculate it. examples - https://github.com/AdguardTeam/tsurlfilter/tree/epic/tswebextension/packages/tsurlfilter/src/rules/declarative-converter#badfilter

Rules with `$cookie` modifier are not supported by AdGuard Content Blocker, AdGuard for iOS, and AdGuard for Safari.

> In [AdGuard Browser Extension for Chrome MV3](/general/adguard-browser-extension/mv3-version) we delete cookies in 2 ways: from `content-script` side (to which we have access) and from `onBeforeSendHeaders` listener. Since `onBeforeSendHeaders` and other listeners are no longer blocking, we are not able to delete them in all cases. You can check if a rule works with [this test](https://testcases.agrd.dev/Filters/cookie-rules/test-cookie-rules).

Reviewer 2: Here and in all other places, it will be more consistent with the document to use `[AdGuard for Chrome MV3][ext-mv3]` and only from that section you link the MV3 page.

Rules with `$cookie` modifier are not supported by AdGuard Content Blocker, AdGuard for iOS, and AdGuard for Safari.

> In AdGuard Browser Extension MV3, the work of `$cookie` rules with parameters is done on the content script side. But since we can no longer hang a blocking listener on events, all of these actions happen asynchronously, and therefore not all changes to cookies will have time to be applied before they are sent to the server.

Reviewer 2: The wording is not ideal.

@105th what did you try to tell us here?

Reviewer 5: @ameshkov we are trying to delete cookies from 2 sides: from content-script (for which we have access) and from onBeforeSendHeaders listener. but since `onBeforeSendHeaders` and other listeners are not blocking anymore we have some race condition here: sometimes we managed to delete them before sending, sometimes not.

it can be checked with this test - https://testcases.agrd.dev/Filters/cookie-rules/test-cookie-rules.
first test here (blocks request cookie) sometimes fails, sometimes not.

- Rules with `$redirect` modifier are not supported by AdGuard Content Blocker, AdGuard for iOS and AdGuard for Safari.
- `$redirect` in uBlock Origin supports specifying priority, e.g. `$redirect=noopjs:42`. AdGuard does not support it and instead just discards the priority postfix.

> - In AdGuard Browser Extension MV3 extension, allowlist rules with `$redirect` are not supported.

Reviewer 2: @105th all of them or is there a case where they work?

Reviewer 5: currently all of them :(

:::info Compatibility

> Rules with `$redirect-rule` modifier are not supported by AdGuard Content Blocker, AdGuard for iOS, AdGuard for Safari, and AdGuard Browser Extension MV3.

Reviewer 2: @105th it makes sense to link the discussion in the W3C group.

Reviewer 5: @SeverAnna https://github.com/w3c/webextensions/issues/493

:::info Compatibility

Rules with the `$referrerpolicy` modifier are supported by AdGuard for Windows, Mac, and Android with [CoreLibs] v1.12 or later.

> In AdGuard Browser Extension MV3, rules with the `$referrerpolicy` modifier are supported but not yet implemented.

Reviewer 2: Argh

Reviewer 5: @SeverAnna we will add it

:::

:::caution Restrictions

[AdGuard Browser Extension for Chrome MV3](/general/adguard-browser-extension/mv3-version) has some limitations:

- Regular expressions, negation and allowlist rules are not supported.
- Group of similar `$removeparam` rules will be combined into one. Example:

> ```bash

Reviewer 2: The formatting seems to be incorrect here and below.

If you want the content to be on the same level as the list item, you need to pad it with 4 spaces.

:::

:::caution Restrictions

AdGuard Browser Extension MV3 has some limitations:

- Allowlist rules are not supported.

> - Regexp, negation and allow rules are not supported.

Reviewer 2: - Allowlist rules are not supported
- allow rules are not supported

Isn't it the same?

> - Regexp, negation and allow rules are not supported.

Reviewer 2: Regexp -> Regular expressions

> - Rules with the same matching condition are combined into one, but only within the scope of one static filter or within the scope of all dynamic rules (custom filters and user rules).

Reviewer 2: It's impossible to understand what it means

Reviewer 5: @SeverAnna we can add examples here

:::info Compatibility

> Such rules with `$app` modifier are supported by AdGuard for Windows, AdGuard for Mac, AdGuard for and Android.

Reviewer 3: ```suggestion
Such rules with `$app` modifier are supported by AdGuard for Windows, AdGuard for Mac, and AdGuard for Android.
```

Reviewer 3: or
```suggestion
Such rules with `$app` modifier are supported by AdGuard for Windows, Mac, and Android.
```
like in the strings above

But rules with mixed style domains restriction are considered invalid. So, for example, the rule
`[$domain=example.org]example.com##.textad` will be ignored.

### Non-basic `$domain` modifier limitations {#non-basic-domain-modifier-limitations}

:::caution Limitations

Since the non-basic `$domain` works the same as the basic one,
it has the same [limitations](#domain-modifier-limitations).

:::

:::info Compatibility

Such rules with `$domain` modifier are supported by AdGuard for Windows, AdGuard for Mac, AdGuard for Android,
AdGuard Browser Extension for Chrome, Firefox, Edge,

> and AdGuard Chrome for MV3.

Reviewer 3: AdGuard Chrome for MV3 is inconsistent with other product names
is it final?

Author: It is not final, I will change it when the PMs have a decision on a name

> 1. `CoreLibs apps` — AdGuard for Windows, AdGuard for Mac, and AdGuard for Android

Reviewer 2: Link corresponding KB pages for every product

1. `AdGuard for Chromium` — AdGuard Browser Extension for Chrome and other Chromium-based browsers such as Microsoft Edge, Opera

> 1. `AdGuard for Chrome MV3` — AdGuard Browser Extension for Chrome MV3

Reviewer 2: Link `/general/adguard-browser-extension/mv3-version`
